Cython/Compiler/ParseTreeTransforms.py

```diff
@@ -3114,6 +3115,21 @@ def _inject_eval(self, node, func_name):
                     node.pos, self.current_scope_node(), lenv))
         return node
 
+    def _inject_class(self, node):
+        if node.name == u'__class__':
+            # bare __class__ reference inside function
+            def_node = self.current_scope_node()
+            if (not isinstance(def_node, Nodes.DefNode) or not def_node.args or
+                len(self.env_stack) < 2):
+                return node
+            class_node, class_scope = self.env_stack[-2]
```

Comment:
This is a bit of an assumption. :)
Instead, you can follow the `parent_scope` chain of the current scope to find the outer class.

Reply:
I just realised that the current `super()` implementation uses exactly this code. Probably worth at least extracting it into a common method to make it easier to change later.